File: server/internal/gamify/catalogue.go
package gamify

// Achievement tiers (docs/SPEC.md XP sources, defaults — tune in alpha):
// paid once, on the day the achievement is earned.
const (
	XpEasy   = 100
	XpMedium = 250
	XpHard   = 500
)

// Achievement keys — the ledger's and the client's vocabulary.
const (
	keySunrise    = "sunrise-club"
	keyNightShift = "night-shift"
	key200Rides   = "200-rides"
	keySufferfest = "sufferfest-survivor"
	keyHotEnd     = "hot-end"
	keyEspresso   = "espresso-ride"
	keyLounge     = "lounge-lizard"
	keyDJ         = "dj"
	keyCrewChief  = "crew-chief"
	keySprintSnob = "sprint-snob"
)

// Achievement is one catalogue entry. The client's copy of this table lives
// in web/src/lib/trophies/catalogue.ts — catalogue_test.go keeps the two in
// step by key, in order. The words a rider reads, the name and the rule, are
// the client's alone (#2876 L9-10): the server never sent them, and a second
// copy here had already drifted from what riders saw.
type Achievement struct {
	Key string
	// The lucide icon name the client draws, like room icons (#447).
	Icon string
	XP   int
	// How many of the counted thing earn it. Zero for the ride
	// achievements, which are judged per ride at save time from the
	// samples in hand and show no partial progress.
	Need int
}

// Catalogue is every achievement the server can verify on its own (#467).
// The Quiet Type (never unmuting) and Never Gonna Give You Up (a track
// queued "as a joke") are not here: mute state is client-reported and a
// joke is not a fact the server holds.
var Catalogue = []Achievement{
	{Key: keySunrise, Icon: "sunrise", XP: XpEasy, Need: 5},
	{Key: keyNightShift, Icon: "moon", XP: XpEasy, Need: 5},
	{Key: key200Rides, Icon: "bike", XP: XpHard, Need: 200},
	{Key: keySufferfest, Icon: "skull", XP: XpHard},
	{Key: keyHotEnd, Icon: "flame", XP: XpMedium},
	{Key: keyEspresso, Icon: "coffee", XP: XpMedium},
	{Key: keyLounge, Icon: "headphones", XP: XpMedium, Need: 10 * 60},
	{Key: keyDJ, Icon: "music", XP: XpMedium, Need: 50},
	{Key: keyCrewChief, Icon: "users", XP: XpHard, Need: 20},
	{Key: keySprintSnob, Icon: "zap", XP: XpMedium, Need: 10},
}

func byKey(key string) (Achievement, bool) {
	for _, a := range Catalogue {
		if a.Key == key {
			return a, true
		}
	}
	return Achievement{}, false
}

File: web/src/lib/trophies/catalogue.ts
import {
	Bike,
	Coffee,
	Flame,
	Headphones,
	Moon,
	Music,
	Skull,
	Sunrise,
	Users,
	Zap,
} from '@lucide/svelte';
import type { Icon } from '$lib/icons';

/**
 * The achievement catalogue (#467) — the client's copy of
 * server/internal/gamify/catalogue.go. The server's catalogue_test.go holds
 * the two to the same keys, icons and XP, in the same order; the server
 * decides who earned what, this file only knows how to draw it. Every
 * number is docs/SPEC.md's.
 */
export interface AchievementMeta {
	key: string;
	name: string;
	how: string;
	icon: Icon;
	xp: number;
	/** What a progress count is in, when it is not simply a count. */
	unit?: string;
}

export const ACHIEVEMENTS: AchievementMeta[] = [
	{
		key: 'sunrise-club',
		name: 'Sunrise Club',
		how: '5 rides started before 07:00, server time',
		icon: Sunrise,
		xp: 100,
	},
	{
		key: 'night-shift',
		name: 'Night Shift',
		how: '5 rides ended after 23:00, server time',
		icon: Moon,
		xp: 100,
	},
	{
		key: '200-rides',
		name: '200 Rides',
		how: 'Two hundred rides on WattRoom',
		icon: Bike,
		xp: 500,
	},
	{
		key: 'sufferfest-survivor',
		name: 'Sufferfest Survivor',
		how: '45 minutes at or above FTP in one ride',
		icon: Skull,
		xp: 500,
	},
	{
		key: 'hot-end',
		name: 'Hot End',
		how: '3 minutes in Z6 in one ride',
		icon: Flame,
		xp: 250,
	},
	{
		key: 'espresso-ride',
		name: 'Espresso Ride',
		how: 'A ride under 25 minutes, 80 % of it above sweet spot',
		icon: Coffee,
		xp: 250,
	},
	{
		key: 'lounge-lizard',
		name: 'Lounge Lizard',
		how: '10 hours in a lounge, in voice',
		icon: Headphones,
		xp: 250,
		unit: 'min',
	},
	{
		key: 'dj',
		name: 'DJ',
		how: 'Queue 50 tracks the room played to the end',
		icon: Music,
		xp: 250,
	},
	{
		key: 'crew-chief',
		name: 'Crew Chief',
		how: 'Coach 20 sessions with 3+ riders',
		icon: Users,
		xp: 500,
	},
	{
		key: 'sprint-snob',
		name: 'Sprint Snob',
		how: 'Win 10 sprint moments on w/kg',
		icon: Zap,
		xp: 250,
	},
];

export const ACHIEVEMENT_BY_KEY: ReadonlyMap<string, AchievementMeta> = new Map(
	ACHIEVEMENTS.map((a) => [a.key, a]),
);
